1_19_R2/src/main/java/me/gamercoder215/mobchip/abstraction/ChipUtil1_19_R2.java

```diff
@@ -0,0 +1,1565 @@
+            return custom(g);
+        }
+    }
+
+    public static ItemStack fromNMS(net.minecraft.world.item.ItemStack item) { return CraftItemStack.asBukkitCopy(item); }
+
+    public static <T> void changeRegistryLock(Registry<T> r, boolean isLocked) {
+        DedicatedServer srv = ((CraftServer) Bukkit.getServer()).getServer();
+        MappedRegistry<T> registry = (MappedRegistry<T>) srv.registryAccess().registryOrThrow(r.key());
+        try {
+            Field frozen = registry.getClass().getDeclaredField("ca");
+            frozen.setAccessible(true);
+            frozen.set(registry, isLocked);
+        } catch (Exception e) {
+            Bukkit.getLogger().severe(e.getClass().getSimpleName());
+            Bukkit.getLogger().severe(e.getMessage());
+            for (StackTraceElement s : e.getStackTrace()) Bukkit.getLogger().severe(s.toString());
+        }
+    }
+
+    @Override
+    public Attribute registerAttribute(NamespacedKey key, double defaultV, double min, double max, boolean client) {
+        if (existsAttribute(key)) return null;
+        changeRegistryLock(BuiltInRegistries.ATTRIBUTE, false);
+
+        DedicatedServer server = ((CraftServer) Bukkit.getServer()).getServer();
+        WritableRegistry<net.minecraft.world.entity.ai.attributes.Attribute> writable = (WritableRegistry<net.minecraft.world.entity.ai.attributes.Attribute>) server.registryAccess().registryOrThrow(Registries.ATTRIBUTE);
+        ResourceKey<net.minecraft.world.entity.ai.attributes.Attribute> nmsKey = ResourceKey.create(Registries.ATTRIBUTE, toNMS(key));
+        Attribute1_19_R2 att = new Attribute1_19_R2(key, defaultV, min, max, client);
+        writable.register(nmsKey, att, Lifecycle.stable());
+
+        changeRegistryLock(BuiltInRegistries.ATTRIBUTE, true);
+        return att;
+    }
+
+    @Override
+    public boolean existsAttribute(NamespacedKey key) {
+        return BuiltInRegistries.ATTRIBUTE.containsKey(toNMS(key));
+    }
+
+    public static ResourceLocation toNMS(NamespacedKey key) {
+        return CraftNamespacedKey.toMinecraft(key);
+    }
+
+    @Override
+    public Attribute getAttribute(NamespacedKey key) {
+        net.minecraft.world.entity.ai.attributes.Attribute a = BuiltInRegistries.ATTRIBUTE.get(toNMS(key));
+        if (!(a instanceof RangedAttribute)) return null;
+        return new Attribute1_19_R2((RangedAttribute) a);
+    }
+
+    @Override
+    public AttributeInstance getAttributeInstance(Mob m, Attribute a) {
+        net.minecraft.world.entity.ai.attributes.Attribute nmsAttribute = BuiltInRegistries.ATTRIBUTE.get(toNMS(a.getKey()));
+        return new AttributeInstance1_19_R2(a, toNMS(m).getAttribute(nmsAttribute));
+    }
+
+    public static net.minecraft.world.entity.ai.gossip.GossipType toNMS(GossipType t) {
+        return net.minecraft.world.entity.ai.gossip.GossipType.valueOf(t.getKey().getKey().toUpperCase());
```

Comment:
`GossipType.byId(String)` is gone, but the class now extends [Codec](https://kvverti.github.io/Documented-DataFixerUpper/snapshot/com/mojang/serialization/Codec.html). I assume Codec has the same functionality that `byId` did, but I can't figure out how to use it, so I've replaced it with `valueOf`.

Reply:
I didn't use `valueOf` because the enum name will obfuscate on runtime (because of the remapping plugin). I'll take a look at the source code when I get the chance.